src/azure-cli-tests/azure/cli/tests/base.py

```diff
@@ -0,0 +1,167 @@
+        recordings_dir = os.path.join(os.path.dirname(test_file_path), 'recordings')
+        live_test = os.environ.get(ENV_LIVE_TEST, None) == 'True'
+
+        self.vcr = vcr.VCR(
+            cassette_library_dir=recordings_dir,
+            before_record_request=self._process_request_recording,
+            before_record_response=self._process_response_recording,
+            decode_compressed_response=True,
+            record_mode='once' if not live_test else 'all',
+            filter_headers=['authorization'],
+        )
+
+        self.recording_file = os.path.join(recordings_dir, '{}.yaml'.format(method_name))
+        if live_test and os.path.exists(self.recording_file):
+            os.remove(self.recording_file)
+
+        self.skip_assert = os.environ.get(ENV_SKIP_ASSERT, None) == 'True'
+        self.in_recording = live_test or not os.path.exists(self.recording_file)
+        self.test_resources_count = 0
+
+    def setUp(self):
+        super(ScenarioTest, self).setUp()
+
+        # set up cassette
+        cm = self.vcr.use_cassette(self.recording_file)
+        self.cassette = cm.__enter__()
+        self.addCleanup(cm.__exit__)
+
+        # set up mock patches
+        patch_main_exception_handler(self)
+        if not self.in_recording:
+            patch_load_cached_subscriptions(self)
+            patch_retrieve_token_for_user(self)
+
+    def create_random_name(self, prefix, length):
+        self.test_resources_count += 1
+        moniker = '{}{:06}'.format(prefix, self.test_resources_count)
+
+        if self.in_recording:
+            name = create_random_name(prefix, length)
+            self.name_replacer.register_name_pair(name, moniker)
+            return name
+        else:
+            return moniker
+
+    def cmd(self, command, checks=None):
+        result = execute(command)
+
+        if not checks:
+            checks = []
+        elif not isinstance(checks, list):
+            checks = [checks]
+
+        if not self.skip_assert:
+            for c in checks:
+                c(result)
+
+        return result
+
+    def _process_request_recording(self, request):
+        for processor in self.recording_processors:
+            request = processor.process_request(request)
+            if not request:
+                break
+
+        return request
+
+    def _process_response_recording(self, response):
+        for processor in self.recording_processors:
+            response = processor.process_response(response)
+            if not response:
+                break
+
+        return response
+
+
+class ExecutionResult(object):
+    def __init__(self, command, expect_failure=False, in_process=True):
+        if in_process:
+            self._in_process_execute(command)
+        else:
+            self._out_of_process_execute(command)
+
+        if expect_failure and self.exit_code == 0:
+            raise AssertionError('The command is expected to fail but it doesn\'.')
```

Comment:
The old version wasn't exactly great either--I think it would "allow exception" but not insist on them... essentially you could just wrap the statement in `with self.assertRaises(...):` but the problem is that EVERYTHING ends up being a CLIError so checking the message was the most reliable way to test the actual error.